### PR TITLE
wallet: Postpone wallet loading notification for encrypted wallets

### LINKED ISSUE #571: Bech32m unavailable for encrypted wallets
bitcoin-core/gui#459 introduced an ability to chose Bech32m address type in the GUI.

On master and on v23.0rc2 this option is unavailable for a just created encrypted wallet, at least on Ubuntu 22.04.

For macOS a slightly different behavior [reported](https://github.com/bitcoin/bitcoin/issues/24501#issuecomment-1081621502).

See details:
- https://github.com/bitcoin/bitcoin/issues/24501#issuecomment-1080821599
- https://github.com/bitcoin/bitcoin/issues/24501#issuecomment-1081598045

### PR DESCRIPTION
**Pull request by hebasto**
Fixes bitcoin-core/gui#571.

`CWallet::Create()` notifies about wallet loading too early, that results the notification goes before `DescriptorScriptPubKeyMan`s were created and added to an encrypted wallet.

And `interfaces::Wallet::taprootEnabled()` in https://github.com/bitcoin/bitcoin/blob/ecf692b466860f44334a1da967fc2559da913bec/src/qt/receivecoinsdialog.cpp#L100-L102 erroneously returns `false` for just created encrypted descriptor wallets.

### CODE REVIEW

File: src/wallet/test/wallet_tests.cpp [hunk 1]
```diff
@@ -778,34 +778,6 @@ BOOST_FIXTURE_TEST_CASE(CreateWallet, TestChain100Setup)
     BOOST_CHECK_EQUAL(addtx_count, 4);
 
 
-    TestUnloadWallet(std::move(wallet));
```
Comment (Sjors) [inline src/wallet/test/wallet_tests.cpp:781]: 471da60a5f16959480be99c77b995165b585c392  Why are you deleting this part of the test?

**Review by Sjors — CHANGES_REQUESTED** [review]
tACK bf01a515f79d36577bd09986ff002ad6296e431f modulo mysteriously disappearing test

File: src/wallet/test/wallet_tests.cpp (hunk 1 above)
Comment (hebasto) [reply src/wallet/test/wallet_tests.cpp:781]: It got broken. Cannot point out the reason though.

Happy to accept any help with this test.

Comment (Sjors) [reply src/wallet/test/wallet_tests.cpp:781]: It was written and later refactored by @ryanofsky.

**Comment by achow101** [comment]
The test fails because `TestLoadWallet` uses `CWallet::Create` directly. As that function no longer calls the wallet loading functions, the things that are supposed to happen in the test case don't happen and causes it to fail.

**Comment by achow101** [comment]
Another alternative is to just put `NotifyWalletLoaded` in `TestLoadWallet`.

**Comment by hebasto** [comment]
> Another alternative is to just put `NotifyWalletLoaded` in `TestLoadWallet`.

```
error: in "wallet_tests/CreateWallet": check addtx_count == 4 has failed [3 != 4]
```

**Comment by achow101** [comment]
This diff works for me:

```diff
diff --git a/src/wallet/test/wallet_tests.cpp b/src/wallet/test/wallet_tests.cpp
index efb19518d0..683f0eb327 100644
--- a/src/wallet/test/wallet_tests.cpp
+++ b/src/wallet/test/wallet_tests.cpp
@@ -54,6 +54,7 @@ static const std::shared_ptr<CWallet> TestLoadWallet(WalletContext& context)
     std::vector<bilingual_str> warnings;
     auto database = MakeWalletDatabase("", options, status, error);
     auto wallet = CWallet::Create(context, "", std::move(database), options.create_flags, error, warnings);
+    NotifyWalletLoaded(context, wallet);
     if (context.chain) {
         wallet->postInitProcess();
     }
@@ -778,6 +779,34 @@ BOOST_FIXTURE_TEST_CASE(CreateWallet, TestChain100Setup)
     BOOST_CHECK_EQUAL(addtx_count, 4);
 
 
+    TestUnloadWallet(std::move(wallet));
+
+
+    // Load wallet again, this time creating new block and mempool transactions
+    // paying to the wallet as the wallet finishes loading and syncing the
+    // queue so the events have to be handled immediately. Releasing the wallet
+    // lock during the sync is a little artificial but is needed to avoid a
+    // deadlock during the sync and simulates a new block notification happening
+    // as soon as possible.
+    addtx_count = 0;
+    auto handler = HandleLoadWallet(context, [&](std::unique_ptr<interfaces::Wallet> wallet) {
+            BOOST_CHECK(rescan_completed);
+            m_coinbase_txns.push_back(CreateAndProcessBlock({}, GetScriptForRawPubKey(coinbaseKey.GetPubKey())).vtx[0]);
+            block_tx = TestSimpleSpend(*m_coinbase_txns[2], 0, coinbaseKey, GetScriptForRawPubKey(key.GetPubKey()));
+            m_coinbase_txns.push_back(CreateAndProcessBlock({block_tx}, GetScriptForRawPubKey(coinbaseKey.GetPubKey())).vtx[0]);
+            mempool_tx = TestSimpleSpend(*m_coinbase_txns[3], 0, coinbaseKey, GetScriptForRawPubKey(key.GetPubKey()));
+            BOOST_CHECK(m_node.chain->broadcastTransaction(MakeTransactionRef(mempool_tx), DEFAULT_TRANSACTION_MAXFEE, false, error));
+            SyncWithValidationInterfaceQueue();
+        });
+    wallet = TestLoadWallet(context);
+    BOOST_CHECK_EQUAL(addtx_count, 4);
+    {
+        LOCK(wallet->cs_wallet);
+        BOOST_CHECK_EQUAL(wallet->mapWallet.count(block_tx.GetHash()), 1U);
+        BOOST_CHECK_EQUAL(wallet->mapWallet.count(mempool_tx.GetHash()), 1U);
+    }
+
+
     TestUnloadWallet(std::move(wallet));
 }
 
```

**Comment by hebasto** [comment]
@achow101 

> This diff works for me:

Thanks! Updated.

**Comment by MarcoFalke** [comment]
Not sure about the last commit. I think it makes sense to limit the scope of locks. And given that this is a backport bugfix, we might want to minimize the diff.

**Comment by Sjors** [comment]
re-utACK 0c12f0116ca802f55f5ab43e6c4842ac403b9889

**Comment by fanquake** [comment]
Backported in #24725.

**Review by dunxen — COMMENTED** [review]
Post-merge tACK 0c12f01

Got to this when going through v23.0rc3 testing. I've checked against this PR and now it works as expected.

**Pull request merged**